### ISSUE TITLE
training YOLOv3 on GPU - CUDA error

### ISSUE THREAD

**Issue by jennigato** [AlexeyAB/darknet#2555]
I'm trying to train tiny yolov3 on GPU with NViDIA RTX 2080 on Ubantu 18.04. After running

```
./darknet detector train data/football.data cfg/yolov3-tiny-football.cfg yolov3-tiny.conv.15 
```
I get the error message:

```
 Allocate additional workspace_size = 17.84 MB 
Loading weights from yolov3-tiny.conv.15...
 seen 64 
Done!
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
Resizing
512 x 512 
 try to allocate additional workspace_size = 37.75 MB 
 CUDA allocate done! 
Loaded: 0.238517 seconds
CUDA status Error: file: ./src/gemm.c : () : line: 2676 : build time: Mar  7 2019 - 22:05:36 
CUDA Error: invalid device symbol
CUDA Error: invalid device symbol: File exists
darknet: ./src/utils.c:281: error: Assertion `0' failed.
Aborted
```
I initially had CUDA 10.0 so I uncommented ```ARCH= -gencode arch=compute_75,code=[sm_75,compute_75] ```  in my Makefile but this gave me this compilation error:

```
...
tmpxft_00015173_00000000-5_maxpool_layer_kernels.cudafe1.cpp:(.text+0x176): undefined reference to `__cudaPopCallConfiguration'
obj/maxpool_layer_kernels.o: In function `forward_maxpool_layer_gpu':
tmpxft_00015173_00000000-5_maxpool_layer_kernels.cudafe1.cpp:(.text+0x266): undefined reference to `__cudaPushCallConfiguration'
obj/maxpool_layer_kernels.o: In function `__device_stub__Z29backward_maxpool_layer_kerneliiiiiiiPfS_Pi(int, int, int, int, int, int, int, float*, float*, int*)':
tmpxft_00015173_00000000-5_maxpool_layer_kernels.cudafe1.cpp:(.text+0x446): undefined reference to `__cudaPopCallConfiguration'
obj/maxpool_layer_kernels.o: In function `backward_maxpool_layer_gpu':
tmpxft_00015173_00000000-5_maxpool_layer_kernels.cudafe1.cpp:(.text+0x542): undefined reference to `__cudaPushCallConfiguration'
obj/avgpool_layer_kernels.o: In function `__device_stub__Z28forward_avgpool_layer_kerneliiiiPfS_(int, int, int, int, float*, float*)':
tmpxft_000151a8_00000000-5_avgpool_layer_kernels.cudafe1.cpp:(.text+0xfb): undefined reference to `__cudaPopCallConfiguration'
obj/avgpool_layer_kernels.o: In function `forward_avgpool_layer_gpu':
tmpxft_000151a8_00000000-5_avgpool_layer_kernels.cudafe1.cpp:(.text+0x1d6): undefined reference to `__cudaPushCallConfiguration'
obj/avgpool_layer_kernels.o: In function `__device_stub__Z29backward_avgpool_layer_kerneliiiiPfS_(int, int, int, int, float*, float*)':
tmpxft_000151a8_00000000-5_avgpool_layer_kernels.cudafe1.cpp:(.text+0x31b): undefined reference to `__cudaPopCallConfiguration'
obj/avgpool_layer_kernels.o: In function `backward_avgpool_layer_gpu':
tmpxft_000151a8_00000000-5_avgpool_layer_kernels.cudafe1.cpp:(.text+0x3f6): undefined reference to `__cudaPushCallConfiguration'
collect2: error: ld returned 1 exit status
Makefile:136: recipe for target 'darknet' failed
make: *** [darknet] Error 1
```
I changed to CUDA 9.0 (commented out arch 75 line), which removed the compilation error but started giving me the invalid device symbol error.

Running nvidia-smi gives me:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 415.18       Driver Version: 415.18       CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 208...  On   | 00000000:09:00.0 Off |                  N/A |
| 63%   68C    P2   176W / 180W |   5876MiB / 10989MiB |     99%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce RTX 208...  On   | 00000000:42:00.0  On |                  N/A |
| 71%   72C    P2   172W / 180W |   5956MiB / 10989MiB |     99%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     80518      C   ./bminer                                    5865MiB |
|    1      5669      G   /usr/lib/xorg/Xorg                            79MiB |
|    1     80518      C   ./bminer                                    5865MiB |
+-----------------------------------------------------------------------------+
```

I get the idea it's some sort of version mismatch but I can't figure out what to do - any help is appreciated!

**Comment by AlexeyAB** [comment]
@jennigato Hi,

Yes, I think there is inconsistent version of NVCC compiler and your CUDA libraries. 

What can you get by this command? `nvcc --version`

> I changed to CUDA 9.0 (commented out arch 75 line), which removed the compilation error but started giving me the invalid device symbol error.

How did you change CUDA to 9.0?

Did you compile by using `make` or `cmake` ?

**Comment by jennigato** [comment]
I compiled using ```make```, running ```nvcc --version```

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, release 9.0, V9.0.176
```
I'm using CUDA on university resource so I set up a new virtualenv and run  ```./vol/cuda/9.0.176/setup.sh```

**Comment by AlexeyAB** [comment]
> CUDA status Error: file: ./src/gemm.c : () : line: 2676 : build time: Mar  7 2019 - 22:05:36 
> CUDA Error: invalid device symbol
> CUDA Error: invalid device symbol: File exists

It is related to cuBlas: https://github.com/AlexeyAB/darknet/blob/b751bac17505a742f149ada81d75689b5e692cde/src/gemm.c#L2674-L2676

So, or set correct paths to `libcublas.so`.

Or try to compile with cuDNN and set `GPU=1 CUDNN=1 CUDNN_HALF=1` in the Makefile.

**Comment by jennigato** [comment]
How do I set the correct paths to ```libcublas.so```? Thanks

**Comment by Adesun** [comment]
you'd better uninstall both cuda 9 & 10,  install cuda10 after reboot  .
cuda10 has better support for RTXs 
for uninstalling you can refer to this[ link ](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#runfile-uninstallation)

**Comment by henbucuoshanghai** [comment]
i have teh same error   ,how to slove it?

**Comment by henbucuoshanghai** [comment]
libdarknet.a(avgpool_layer_kernels.o): In function `__device_stub__Z29backward_avgpool_layer_kerneliiiiPfS_(int, int, int, int, float*, float*)':
tmpxft_0000619c_00000000-5_avgpool_layer_kernels.compute_61.cudafe1.cpp:(.text+0x2ef): undefined reference to `__cudaPopCallConfiguration'
libdarknet.a(avgpool_layer_kernels.o): In function `backward_avgpool_layer_gpu':
tmpxft_0000619c_00000000-5_avgpool_layer_kernels.compute_61.cudafe1.cpp:(.text+0x3cc): undefined reference to `__cudaPushCallConfiguration'
libdarknet.a(crop_layer_kernels.o): In function `__device_stub__Z19levels_image_kernelPfS_iiiifffff(float*, float*, int, int, int, int, float, float, float, float, float)':
tmpxft_0000608a_00000000-5_crop_layer_kernels.compute_61.cudafe1.cpp:(.text+0x1d9): undefined reference to `__cudaPopCallConfiguration'
libdarknet.a(crop_layer_kernels.o): In function `__device_stub__Z25forward_crop_layer_kernelPfS_iiiiiiiifS_(float*, float*, int, int, int, int, int, int, int, int, float, float*)':
tmpxft_0000608a_00000000-5_crop_layer_kernels.compute_61.cudafe1.cpp:(.text+0x3d8): undefined reference to `__cudaPopCallConfiguration'
libdarknet.a(crop_layer_kernels.o): In function `forward_crop_layer_gpu':
tmpxft_0000608a_00000000-5_crop_layer_kernels.compute_61.cudafe1.cpp:(.text+0x547): undefined reference to `__cudaPushCallConfiguration'
tmpxft_0000608a_00000000-5_crop_layer_kernels.compute_61.cudafe1.cpp:(.text+0x5a1): undefined reference to `__cudaPushCallConfiguration'
libdarknet.a(dropout_layer_kernels.o): In function `__device_stub__Z28yoloswag420blazeit360noscopePfiS_ff(float*, int, float*, float, float)':
tmpxft_000060e6_00000000-5_dropout_layer_kernels.compute_61.cudafe1.cpp:(.text+0xd5): undefined reference to `__cudaPopCallConfiguration'
libdarknet.a(dropout_layer_kernels.o): In function `forward_dropout_layer_gpu':
tmpxft_000060e6_00000000-5_dropout_layer_kernels.compute_61.cudafe1.cpp:(.text+0x1da): undefined reference to `__cudaPushCallConfiguration'
libdarknet.a(dropout_layer_kernels.o): In function `backward_dropout_layer_gpu':
tmpxft_000060e6_00000000-5_dropout_layer_kernels.compute_61.cudafe1.cpp:(.text+0x287): undefined reference to `__cudaPushCallConfiguration'
libdarknet.a(maxpool_layer_kernels.o): In function `__device_stub__Z28forward_maxpool_layer_kerneliiiiiiiPfS_Pi(int, int, int, int, int, int, int, float*, float*, int*)':
tmpxft_00006149_00000000-5_maxpool_layer_kernels.compute_61.cudafe1.cpp:(.text+0x16a): undefined reference to `__cudaPopCallConfiguration'
libdarknet.a(maxpool_layer_kernels.o): In function `forward_maxpool_layer_gpu':
tmpxft_00006149_00000000-5_maxpool_layer_kernels.compute_61.cudafe1.cpp:(.text+0x24c): undefined reference to `__cudaPushCallConfiguration'
libdarknet.a(maxpool_layer_kernels.o): In function `__device_stub__Z29backward_maxpool_layer_kerneliiiiiiiPfS_Pi(int, int, int, int, int, int, int, float*, float*, int*)':
tmpxft_00006149_00000000-5_maxpool_layer_kernels.compute_61.cudafe1.cpp:(.text+0x40a): undefined reference to `__cudaPopCallConfiguration'
libdarknet.a(maxpool_layer_kernels.o): In function `backward_maxpool_layer_gpu':
tmpxft_00006149_00000000-5_maxpool_layer_kernels.compute_61.cudafe1.cpp:(.text+0x4f8): undefined reference to `__cudaPushCallConfiguration'
collect2: error: ld returned 1 exit status
Makefile:83: recipe for target 'darknet' failed
make: *** [darknet] Error 1

**Comment by henbucuoshanghai** [comment]
nvcc -V  it is cuda10.0, it is too new?I have to use the cuda9.0?   the nvidai-smi is 418.
i have run many projects with this.
anything wrong?

**Comment by henbucuoshanghai** [comment]
_00000000-5_avgpool_layer_kernels.cudafe1.cpp:(.text+0x3f6): undefined reference to `__cudaPushCallConfiguration'


how to solve it? @jennigato

**Comment by Jmennius** [comment]
The error reposting is broken in this case (see #7417),
the actual error is `CUBLAS_STATUS_EXECUTION_FAILED` when `cublasSgemm()` is invoked.

CUDA Toolkit 11.1 release notes mention an issue fixed in cuBLAS:
> Fixed an issue that caused an Address out of bounds error when calling cublasSgemm().

We had `cublasSgemm()` failing with `CUBLAS_STATUS_EXECUTION_FAILED` for us when built with CUDA Toolkit 10.0 and running on Ampere GPU (3060 Ti). It ran fine on older GPUs (Pascal, Turing).
We had it run successfully on Ampere when we built it with CUDA 11.2.

Basically - try building against the newest CUDA Toolkit (at least 11.1) available and see if it helps.
